pennylane/math/multi_dispatch.py

```diff
@@ -0,0 +1,340 @@
+            "TensorFlow and PyTorch over autograd. Consider replacing Autograd with vanilla NumPy.",
+            UserWarning,
+        )
+
+    if "tensorflow" in interfaces:
+        return "tensorflow"
+
+    if "torch" in interfaces:
+        return "torch"
+
+    if "autograd" in interfaces:
+        return "autograd"
+
+    if "jax" in interfaces:
+        return "jax"
+
+    return "numpy"
+
+
+def block_diag(values):
+    """Combine a sequence of 2D tensors to form a block diagonal tensor.
+
+    Args:
+        values (Sequence[tensor_like]): Sequence of 2D arrays/tensors to form
+            the block diagonal tensor.
+
+    Returns:
+        tensor_like: the block diagonal tensor
+
+    **Example**
+
+    >>> t = [
+    ...     np.array([[1, 2], [3, 4]]),
+    ...     torch.tensor([[1, 2, 3], [-1, -6, -3]]),
+    ...     torch.tensor(5)
+    ... ]
+    >>> qml.math.block_diag(t)
+    tensor([[ 1,  2,  0,  0,  0,  0],
+            [ 3,  4,  0,  0,  0,  0],
+            [ 0,  0,  1,  2,  3,  0],
+            [ 0,  0, -1, -6, -3,  0],
+            [ 0,  0,  0,  0,  0,  5]])
+    """
+    interface = _multi_dispatch(values)
+    values = np.coerce(values, like=interface)
+    return np.block_diag(values, like=interface)
+
+
+def concatenate(values, axis=0):
+    """Concatenate a sequence of tensors along the specified axis.
+
+    .. warning::
+
+        Tensors that are incompatible (such as Torch and TensorFlow tensors)
+        cannot both be present.
+
+    Args:
+        values (Sequence[tensor_like]): Sequence of tensor-like objects to
+            concatenate. The objects must have the same shape, except in the dimension corresponding
+            to axis (the first, by default).
+        axis (int): The axis along which the input tensors are concatenated. If axis is None,
+            tensors are flattened before use. Default is 0.
+
+    Returns:
+        tensor_like: The concatenated tensor.
+
+    **Example**
+
+    >>> x = tf.constant([0.6, 0.1, 0.6])
+    >>> y = tf.Variable([0.1, 0.2, 0.3])
+    >>> z = np.array([5., 8., 101.])
+    >>> concatenate([x, y, z])
+    <tf.Tensor: shape=(3, 3), dtype=float32, numpy=
+    array([6.00e-01, 1.00e-01, 6.00e-01, 1.00e-01, 2.00e-01, 3.00e-01, 5.00e+00, 8.00e+00, 1.01e+02], dtype=float32)>
+    """
+    interface = _multi_dispatch(values)
+
+    if interface == "torch":
+        import torch
+
+        if axis is None:
+            # flatten and then concatenate zero'th dimension
+            # to reproduce numpy's behaviour
+            values = [np.flatten(torch.as_tensor(t)) for t in values]
+            axis = 0
+        else:
+            values = [torch.as_tensor(t) for t in values]
+
+    if interface == "tensorflow" and axis is None:
+        # flatten and then concatenate zero'th dimension
+        # to reproduce numpy's behaviour
+        values = [np.flatten(np.array(t)) for t in values]

```

Comment:
For my understanding, do these NumPy functions work seamlessly on TF objects?

Reply:
Yep :) When you call `autoray.numpy.some_function(...)`, it simply performs single dispatch to determine the correct library to use. E.g:

```python
import numpy as np
import tensorflow as tf
import autoray

x = tf.Variable([0.2, 0.5])
y = np.array([0.1, 0.2])
autoray.numpy.dot(x, y)  # calls tf.matmul(x, y)
autoray.numpy.dot(y, x)  # I believe this will call np.dot(y, x)
```

Reply:
We can register _new_ `autoray.numpy` functions, and dispatch rules, via the `single_dispatch.py` file.